src/trusted/service_runtime/nacl_app_thread.c

```diff
@@ -653,6 +744,17 @@ int NaClAppThreadSpawn(struct NaClAppThread     *natp_parent,
     NaClTlsSetTlsValue2(natp_child, user_tls2);
   }
 
+
+  if (!cage_thread) {
+    natp_child->is_cage_mainthread = true;
+    natp_child->cage_mainthread = NULL;
+    natp_child->tearing_down = false;
+  } 
+  else {
+    natp_child->is_cage_mainthread = false;
+    natp_child->cage_mainthread = natp_parent;
```

Comment:
Wait is this always true?
I'd think you should probably just have the natp for the cage mainthread refer to itself with the cage_mainthread field and then you can set `natp_child->cage_mainthread = natp_parent->cage_mainthread` here.